codebase/src/compiler/checker-inference.md

```diff
@@ -0,0 +1,245 @@
+type is called the **source** type, since it is the source of
+inferences. It's matched with the parameter type `{ initial(): T }`.
+This is the **target** type -- it contains type parameters which are
+the target of the process.
+
+Type parameter inference is a pairwise walk of the two types, looking
+for type parameters in the target, matching them to corresponding
+types in the source. The type is walked structurally sort of like a tree
+is elsewhere in the compiler.
+
+1. `inferTypes` gets called on each source/target pair with
+   argument=source/parameter=target. There's only one pair here:
+   `{ initial(): string }` and `{ initial(): T }`.
+2. Since both sides are object types, `inferFromProperties` looks
+   through each property of the target and looks for a match in the
+   source. In this case both have the property `initial`.
+3. `initial`'s type is a signature on both sides
+   (`() => T/() => string`), so inference goes to `inferFromSignature`, which
+   recursively infers from the return type.
+4. Now the source/target pair is `T/string`. Since the source is a
+   lone type parameters, we add `string` to the list of candidates for
+   `T`.
+
+Once all the parameters have had `inferTypes` called on them,
+`getInferredTypes` condenses each candidate array to a single type,
+via `getUnionType` in this case. `T`'s candidates array is `[string]`,
+so `getUnionType` immediately returns `string`.
+
+### Other considerations
+
+#### Method of Combining Candidate Arrays
+
+Only inference to return types, `keyof T` and mapped type constraints
+(which are usually `keyof` too) produce a union. All other locations
+call the custom code `getCommonSupertype`, which more or less does
+what it says. Note that object types are always unioned together
+first, regardless of inference position.
+
+#### Interference Between Contextual Typing and Type Parameter Inference
+
+Type parameter inference actually operates in two passes. The first
+pass skips arguments that have contextually typed expressions so that
+if good inferences are found from other arguments, contextual typing
+can provide types to parameters of function expressions, which in turn
+may produce better return types. Then the second pass proceeds with
+all arguments.
+
+#### Inference Priorities
+
+Different positions have different inference priorities; when the type
+walk finds a candidate at a higher priority position than existing
+candidates, it throws away the existing candidates and starts over
+with the higher-priority candidate. For example, a lone type variable
+has the highest priority, but a type variable found inside a return type
+has one of the lowest priorities.
+
+#### Contravariant Candidates
+
+Certain candidates are inferred contravariantly, such as parameters of
+callbacks. This is a separate system from inference priorities;
+contravariant candidates are even higher priority.
+
+#### Reverse Mapped Types
```

Comment:
Potentially worth nothing that they are pretty much just as much a first class type as a conditional, an object, or a mapped type - there's just not an (obvious) syntax to directly construct them in the type system - they only arise from inference. This causes us plenty of issues when it comes to serializing them~